File: ClearScript/V8/V8ScriptEngine.InitScript.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.





namespace Microsoft.ClearScript.V8
{
    public sealed partial class V8ScriptEngine
    {
        private const string initScript = "Object.defineProperty(this,'EngineInternal',{value:(e=>{let t=e=>e.bind();function r(){return new this(...arguments)}let o=e.isHostObjectKey;delete e.isHostObjectKey;let n=e=>!!e&&!0===e[o],i=e.moduleResultKey;delete e.moduleResultKey;let c=e.getPromiseState;delete e.getPromiseState;let s=e.getPromiseResult;delete e.getPromiseResult;let a=Promise,l=JSON,u=Symbol(),y=e.toJson;return delete e.toJson,Object.freeze({commandHolder:{},getCommandResult:t(e=>null==e?e:'function'!=typeof e.hasOwnProperty?'Module'===e[Symbol.toStringTag]?'[module]':'[external]':!0===e[o]?e:'function'!=typeof e.toString?'['+typeof e+']':e.toString()),strictEquals:t((e,t)=>e===t),isPromise:t(e=>e instanceof a),isHostObject:t(n),invokeConstructor:t((e,t)=>{if('function'!=typeof e)throw Error('Function expected');return r.apply(e,Array.from(t))}),invokeMethod:t((e,t,r)=>{if('function'!=typeof t)throw Error('Function expected');return t.apply(e,Array.from(r))}),createPromise:t(function(){return new a(...arguments)}),createSettledPromiseWithResult:t(e=>{try{return a.resolve(e())}catch(t){return a.reject(t)}}),createSettledPromise:t(e=>{try{return e(),a.resolve()}catch(t){return a.reject(t)}}),completePromiseWithResult:t((e,t,r)=>{try{t(e())}catch(o){r(o)}}),completePromise:t((e,t,r)=>{try{e(),t()}catch(o){r(o)}}),getPromiseState:t(c),getPromiseResult:t(s),initializeTask:t((e,t,r,o,n)=>{t?e.then(o,n):r?n(s(e)):o(s(e))}),createArray:t(()=>[]),throwValue:t(e=>{throw e}),getStackTrace:t(()=>{try{throw Error('[stack trace]')}catch(e){return e.stack}}),toIterator:t(function*(e){try{for(;e.ScriptableMoveNext();)yield e.ScriptableCurrent}finally{e.ScriptableDispose()}}),toAsyncIterator:t(async function*(e){try{for(;await e.ScriptableMoveNextAsync();)yield e.ScriptableCurrent}finally{await e.ScriptableDisposeAsync()}}),getIterator:t(e=>e?.[Symbol.iterator]?.()),getAsyncIterator:t(e=>e?.[Symbol.asyncIterator]?.()),checkpoint:t(()=>{let t=e[u];if(t)throw t}),toJson:t((e,t)=>y?l.parse(y(e,t)):t),parseJson:t(e=>l.parse(e)),asyncGenerator:async function*(){}().constructor,getModuleResult:t(async(e,t)=>(await e,t[0]?.[i]))})})(this)});";
    }
}

File: ClearScript/V8/V8ScriptEngine.InitScript.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

Object.defineProperty(this, 'EngineInternal', { value: (globalObject => {

    const bind = value => value.bind();

    function construct() {
        return new this(...arguments);
    }

    const isHostObjectKey = globalObject.isHostObjectKey;
    delete globalObject.isHostObjectKey;
    const isHostObject = value => !!value && (value[isHostObjectKey] === true);

    const moduleResultKey = globalObject.moduleResultKey;
    delete globalObject.moduleResultKey;

    const getPromiseState = globalObject.getPromiseState;
    delete globalObject.getPromiseState;

    const getPromiseResult = globalObject.getPromiseResult;
    delete globalObject.getPromiseResult;

    const savedPromise = Promise;
    const savedJSON = JSON;
    const checkpointSymbol = Symbol();

    const toJson = globalObject.toJson;
    delete globalObject.toJson;

    return Object.freeze({

        commandHolder: {},

        getCommandResult: bind(value => {
            if ((value === null) || (value === undefined)) {
                return value;
            }
            if (typeof(value.hasOwnProperty) !== 'function') {
                if (value[Symbol.toStringTag] === 'Module') {
                    return '[module]';
                }
                return '[external]';
            }
            if (value[isHostObjectKey] === true) {
                return value;
            }
            if (typeof(value.toString) !== 'function') {
                return '[' + typeof(value) + ']';
            }
            return value.toString();
        }),

        strictEquals: bind((left, right) => left === right),

        isPromise: bind(value => value instanceof savedPromise),

        isHostObject: bind(isHostObject),

        invokeConstructor: bind((constructor, args) => {
            if (typeof(constructor) !== 'function') {
                throw new Error('Function expected');
            }
            return construct.apply(constructor, Array.from(args));
        }),

        invokeMethod: bind((target, method, args) => {
            if (typeof(method) !== 'function') {
                throw new Error('Function expected');
            }
            return method.apply(target, Array.from(args));
        }),

        createPromise: bind(function () {
            return new savedPromise(...arguments);
        }),

        createSettledPromiseWithResult: bind(getResult => {
            try {
                return savedPromise.resolve(getResult());
            }
            catch (exception) {
                return savedPromise.reject(exception);
            }
        }),

        createSettledPromise: bind(wait => {
            try {
                wait();
                return savedPromise.resolve();
            }
            catch (exception) {
                return savedPromise.reject(exception);
            }
        }),

        completePromiseWithResult: bind((getResult, resolve, reject) => {
            try {
                resolve(getResult());
            }
            catch (exception) {
                reject(exception);
            }
            return undefined;
        }),

        completePromise: bind((wait, resolve, reject) => {
            try {
                wait();
                resolve();
            }
            catch (exception) {
                reject(exception);
            }
            return undefined;
        }),

        getPromiseState: bind(getPromiseState),

        getPromiseResult: bind(getPromiseResult),

        initializeTask: bind((promise, isPending, isRejected, onResolved, onRejected) => {
            if (isPending) {
                promise.then(onResolved, onRejected);
            }
            else if (isRejected) {
                onRejected(getPromiseResult(promise));
            }
            else {
                onResolved(getPromiseResult(promise));
            }
            return undefined;
        }),

        createArray: bind(() => []),

        throwValue: bind(value => {
            throw value;
        }),

        getStackTrace: bind(() => {
            try {
                throw new Error('[stack trace]');
            }
            catch (exception) {
                return exception.stack;
            }
        }),

        toIterator: bind(function* (enumerator) {
            try {
                while (enumerator.ScriptableMoveNext()) {
                    yield enumerator.ScriptableCurrent;
                }
            }
            finally {
                enumerator.ScriptableDispose();
            }
        }),

        toAsyncIterator: bind(async function* (asyncEnumerator) {
            try {
                while (await asyncEnumerator.ScriptableMoveNextAsync()) {
                    yield asyncEnumerator.ScriptableCurrent;
                }
            }
            finally {
                await asyncEnumerator.ScriptableDisposeAsync();
            }
        }),

        getIterator: bind(obj => obj?.[Symbol.iterator]?.()),

        getAsyncIterator: bind(obj => obj?.[Symbol.asyncIterator]?.()),

        checkpoint: bind(() => {
            const value = globalObject[checkpointSymbol];
            if (value) {
                throw value;
            }
        }),

        toJson: bind((key, value) => toJson ? savedJSON.parse(toJson(key, value)) : value),

        parseJson: bind(json => savedJSON.parse(json)),

        asyncGenerator: (async function* () {})().constructor,

        getModuleResult: bind(async (result, metaHolder) => {
            await result;
            return metaHolder[0]?.[moduleResultKey];
        })

    });
})(this) });
